### ISSUE TITLE
can't use bootstrap inference with xlearner

### ISSUE THREAD

**Issue by skfkrh** [microsoft/EconML#333]
I'm trying to fit x_learner and get CI. I can fit it without inference without error, but when I put BootstrapInference, it give the following error.
Did I miss anything?

X_learner.fit(df_avg.outcome, (df_avg.assignment=='treatment').astype(int), df_avg[heteX], 
              inference= BootstrapInference(n_bootstrap_samples=10,n_jobs=-1))

---------------------------------------------------------------------------
Empty                                     Traceback (most recent call last)
~\AppData\Roaming\Python\Python36\site-packages\joblib\parallel.py in dispatch_one_batch(self, iterator)
    807             try:
--> 808                 tasks = self._ready_batches.get(block=False)
    809             except queue.Empty:

~\Anaconda3\lib\queue.py in get(self, block, timeout)
    160                 if not self._qsize():
--> 161                     raise Empty
    162             elif timeout is None:

Empty: 

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-29-f3acf6683c38> in <module>()
     15        'guest_country_H', 'guest_country_I']
     16 X_learner.fit(df_avg.booking_value_usd2, (df_avg.assignment=='treatment').astype(int), df_avg[heteX], 
---> 17               inference= BootstrapInference(n_bootstrap_samples=10,n_jobs=-1))
     18 
     19 

~\AppData\Roaming\Python\Python36\site-packages\econml\cate_estimator.py in call(self, Y, T, inference, *args, **kwargs)
     90             if inference is not None:
     91                 # NOTE: we call inference fit *after* calling the main fit method
---> 92                 inference.fit(self, Y, T, *args, **kwargs)
     93             self._inference = inference
     94             return self

~\AppData\Roaming\Python\Python36\site-packages\econml\inference.py in fit(self, estimator, *args, **kwargs)
     49     def fit(self, estimator, *args, **kwargs):
     50         est = BootstrapEstimator(estimator, self._n_bootstrap_samples, self._n_jobs, compute_means=False)
---> 51         est.fit(*args, **kwargs)
     52         self._est = est
     53 

~\AppData\Roaming\Python\Python36\site-packages\econml\bootstrap.py in fit(self, *args, **named_args)
     75                          *[convertArg(arg, inds) for arg in args],
     76                          **{arg: convertArg(named_args[arg], inds) for arg in named_args})
---> 77             for obj, inds in zip(self._instances, indices)
     78         )
     79         return self

~\AppData\Roaming\Python\Python36\site-packages\joblib\parallel.py in __call__(self, iterable)
   1027             # remaining jobs.
   1028             self._iterating = False
-> 1029             if self.dispatch_one_batch(iterator):
   1030                 self._iterating = self._original_iterator is not None
   1031 

~\AppData\Roaming\Python\Python36\site-packages\joblib\parallel.py in dispatch_one_batch(self, iterator)
    817                 big_batch_size = batch_size * n_jobs
    818 
--> 819                 islice = list(itertools.islice(iterator, big_batch_size))
    820                 if len(islice) == 0:
    821                     return False

~\AppData\Roaming\Python\Python36\site-packages\econml\bootstrap.py in <genexpr>(.0)
     75                          *[convertArg(arg, inds) for arg in args],
     76                          **{arg: convertArg(named_args[arg], inds) for arg in named_args})
---> 77             for obj, inds in zip(self._instances, indices)
     78         )
     79         return self

~\AppData\Roaming\Python\Python36\site-packages\econml\bootstrap.py in <listcomp>(.0)
     73         self._instances = Parallel(n_jobs=self._n_jobs, prefer='threads', verbose=3)(
     74             delayed(fit)(obj,
---> 75                          *[convertArg(arg, inds) for arg in args],
     76                          **{arg: convertArg(named_args[arg], inds) for arg in named_args})
     77             for obj, inds in zip(self._instances, indices)

~\AppData\Roaming\Python\Python36\site-packages\econml\bootstrap.py in convertArg(arg, inds)
     70 
     71         def convertArg(arg, inds):
---> 72             return arg[inds] if arg is not None else None
     73         self._instances = Parallel(n_jobs=self._n_jobs, prefer='threads', verbose=3)(
     74             delayed(fit)(obj,

~\AppData\Roaming\Python\Python36\site-packages\pandas\core\frame.py in __getitem__(self, key)
   2906             if is_iterator(key):
   2907                 key = list(key)
-> 2908             indexer = self.loc._get_listlike_indexer(key, axis=1, raise_missing=True)[1]
   2909 
   2910         # take() does not accept boolean indexers

~\AppData\Roaming\Python\Python36\site-packages\pandas\core\indexing.py in _get_listlike_indexer(self, key, axis, raise_missing)
   1252             keyarr, indexer, new_indexer = ax._reindex_non_unique(keyarr)
   1253 
-> 1254         self._validate_read_indexer(keyarr, indexer, axis, raise_missing=raise_missing)
   1255         return keyarr, indexer
   1256 

~\AppData\Roaming\Python\Python36\site-packages\pandas\core\indexing.py in _validate_read_indexer(self, key, indexer, axis, raise_missing)
   1296             if missing == len(indexer):
   1297                 axis_name = self.obj._get_axis_name(axis)
-> 1298                 raise KeyError(f"None of [{key}] are in the [{axis_name}]")
   1299 
   1300             # We (temporarily) allow for some missing keys with .loc, except in

KeyError: "None of [Int64Index([405999,  68319, 250946, 398394,   8195, 174595, 266031, 114482,\n             24971,  96134,\n            ...\n            175332, 153775, 114614, 120308, 213296,  79010,  46927, 389005,\n            151523,  91304],\n           dtype='int64', length=410035)] are in the [columns]"

**Comment by skfkrh** [comment]
Also additional question is that in the text it's noted that metalearner doesn't provide theoretically sound confidence interval. Is this also true for well balanced data sets?

**Comment by vsyrgkanis** [comment]
@skfkrh which version are you using?
This seems related to the fact that you are passing pandas data frames to the package. For more robust pass numpy arrays.

Though @moprescu recently pushed a fix that allows us to accept pandas. So this shouldn't come up in the latest release (8.1).

@moprescu can you also check this error?

**Comment by skfkrh** [comment]
Yes this was pandas issues. I have older version of econml. thank you.
Could you let me know the answer for my second comment?
